### PR TITLE
daemon: Sync local IPs to lxcmap periodically

### LINKED ISSUE #3870: Host IP disappears from BPF endpoint map
We have seen instances where the IP assigned to the `cilium_host` interface disappears from the endpoint list when Cilium is getting restarted frequently.

The most likely explanation is the following:
1. Cilium start allocates an IP out of the address range for cilium_host
2. Endpoint restore happens and a previous endpoint was using the IP that was allocated for cilium_host in the previous step
3. The endpoint generation/restore overwrite the cilium_host entry.

### PR DESCRIPTION
**Pull request by jrajahalme**
LXCMap should not get out of sync, but there is some evidence that
sometimes it does. Add a new controller to refresh the host entries in
the lxcmap every 5 seconds, but only if they are not already there.

No garbage collection of potentially stale host entries in the lxcmap
is done.

lxcmap implementation of bpf.MapValue is fixed to use a pointer receiver
so that lookups function as intended.

Fixes: #3870

### CODE REVIEW

**Comment by jrajahalme** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]

**Comment by jrajahalme** [comment]
test-me-please

File: daemon/daemon.go [hunk 1]
```diff
@@ -905,6 +883,51 @@ func (d *Daemon) init() error {
...
+					id, exists := ipcache.IPIdentityCache.LookupByIP(ip.String())
+					if !exists || id != identity.ReservedIdentityHost {
+						// Upsert will not propagate (reserved:host->ID) mappings across the cluster.
+						// Manually push it into each IPIdentityMappingListener.
+						log.WithField(logfields.IPAddr, ip).Debug("Adding local ip to ipcache")
+						ipcache.IPIdentityCache.Upsert(ip.String(), identity.ReservedIdentityHost)
+
+						localIPIDPair := identity.IPIdentityPair{IP: ip, ID: identity.ReservedIdentityHost}
+						d.OnIPIdentityCacheChange(ipcache.Upsert, localIPIDPair)
+						envoy.NetworkPolicyHostsCache.OnIPIdentityCacheChange(ipcache.Upsert, localIPIDPair)
```
Comment (joestringer) [inline daemon/daemon.go]: I believe that this will cause problems if we ever do this insert twice.

EDIT: Which *shouldn't* happen if the `ipcache.IPIdentityCache.LookupByIP()` logic works as expected. Should be fine, I think.

File: daemon/daemon.go [hunk 2]
```diff
@@ -905,6 +883,51 @@ func (d *Daemon) init() error {
...
+	if err := lxcmap.LXCMap.DeleteAll(); err != nil {
+		return err
+	}
+
+	controller.NewManager().UpdateController("lxcmap-bpf-host-sync",
+		controller.ControllerParams{
+			DoFunc: func() error {
+				// TODO: Update addresses first, in case node addressing has changed.
+				// TODO: Once these start changing on runtime, figure out the locking strategy.
+				// TODO: Delete stale host entries from the lxcmap.
```
Comment (joestringer) [inline daemon/daemon.go]: If you're planning to bind lxcmap update to ipcache update, then ipcache update already has GC, so maybe you could put the GC there..? (See `OnIPIdentityCacheGC()`).

**Review by joestringer — COMMENTED** [review]
It's not obvious in your commit message, but this is performing ipcache sync as well. This has some implications. I would suggest running this with short timers and L7 policies, and keep an eye on things like the logs to see that Cilium is operating sanely.

File: pkg/maps/lxcmap/lxcmap.go [hunk 3]
```diff
@@ -192,10 +193,20 @@ func WriteEndpoint(f EndpointFrontend) error {
...
-	ep := EndpointInfo{Flags: EndpointFlagHost}
+	ep := &EndpointInfo{Flags: EndpointFlagHost}
 	return LXCMap.Update(key, ep)
 }
 
+func SyncHostEntry(ip net.IP) error {
+	key := NewEndpointKey(ip)
+	value, err := LXCMap.Lookup(key)
+	if err != nil || value.(*EndpointInfo).Flags&EndpointFlagHost == 0 {
+		log.WithError(err).WithField(logfields.IPAddr, ip).Debugf("Adding local ip to endpoint map")
```
Comment (joestringer) [inline pkg/maps/lxcmap/lxcmap.go]: There's nothing that is specific about this being a `local ip` here, perhaps we can shift this to the caller.

File: daemon/daemon.go [hunk 4]
```diff
@@ -905,6 +883,51 @@ func (d *Daemon) init() error {
...
+					node.GetIPv6(),
+					node.GetIPv6Router(),
+				}
+
+				for _, ip := range localIPs {
+					if err := lxcmap.SyncHostEntry(ip); err != nil {
+						return fmt.Errorf("Unable to add host entry to endpoint map: %s", err)
+					}
+					id, exists := ipcache.IPIdentityCache.LookupByIP(ip.String())
+					if !exists || id != identity.ReservedIdentityHost {
```
Comment (joestringer) [inline daemon/daemon.go]: If the ipcache thinks that the identity for the IP is not `reserved:host`, then this will trigger. Then presumably re-trigger, again and again as long as the entry exists in the ipcache?

File: pkg/bpf/endpoint.go [hunk 5]
```diff
@@ -40,7 +40,7 @@ type EndpointKey struct {
 // GetKeyPtr returns the unsafe pointer to the BPF key
 func (k EndpointKey) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(&k) }
 
-// GetValuePtr returns the unsafe pointer to the BPF value
+// GetValuePtr returns the unsafe pointer to the BPF KEY <<<< NOTE!
```
Comment (tgraf) [inline pkg/bpf/endpoint.go]: Please move this commit to a separate PR. It needs improvements but I think fixing the comment like this is not what we want because it was still not clear to me even after this comment ;-)

Let's move it to a separate PR as we won't have to backport this.

File: pkg/maps/lxcmap/lxcmap.go (hunk 3 above)
Comment (tgraf) [reply pkg/maps/lxcmap/lxcmap.go]: I would prefer to return an error here and log in the caller

File: daemon/daemon.go [hunk 6]
```diff
@@ -905,6 +883,51 @@ func (d *Daemon) init() error {
...
+						localIPIDPair := identity.IPIdentityPair{IP: ip, ID: identity.ReservedIdentityHost}
+						d.OnIPIdentityCacheChange(ipcache.Upsert, localIPIDPair)
+						envoy.NetworkPolicyHostsCache.OnIPIdentityCacheChange(ipcache.Upsert, localIPIDPair)
+					}
+				}
+				return nil
+			},
+			RunInterval: time.Duration(5) * time.Second,
+		})
+	// TODO: To keep the old behavior we should wait for one successful completion of the controller
```
Comment (tgraf) [inline daemon/daemon.go]: We can't keep this as a todo as we might get into a situation where no host IP are synced down.

You can do the following:
- extract the controller run function into a separate function and run it once and fail on any error to keep behaviour on failing while bootstrapping when any of the host IPs can't be added.
- or create channel and pass it into the controller and then close it on the first successful run. You could then wait on the channel to be closed outside with a timeout. This would give us retries even on bootstrapping but the logic is a bit more complicated.

**Review by tgraf — CHANGES_REQUESTED** [review]
Please submit patch 3 separately as it fixes a separate bug as well. We are doing backports on PR level so for fixes that we target for backports, we should try to isolate PRs as much as possible.

File: daemon/daemon.go [hunk 7]
```diff
@@ -905,6 +883,51 @@ func (d *Daemon) init() error {
...
+				}
+
+				for _, ip := range localIPs {
+					if err := lxcmap.SyncHostEntry(ip); err != nil {
+						return fmt.Errorf("Unable to add host entry to endpoint map: %s", err)
+					}
+					id, exists := ipcache.IPIdentityCache.LookupByIP(ip.String())
+					if !exists || id != identity.ReservedIdentityHost {
+						// Upsert will not propagate (reserved:host->ID) mappings across the cluster.
+						// Manually push it into each IPIdentityMappingListener.
```
Comment (raybejjani) [inline daemon/daemon.go]: @joestringer has a PR to add a UpdateToKVStore. It does propagate to the listeners via the KVStore. If we use that this could then be:
```golang
ipcache.IPIdentityCache.Upsert(ip.String(), identity.ReservedIdentityHost)
ipcache.IPIdentityCache.UpsertToKVStore(ip.String(), identity.ReservedIdentityHost) <-- I don't remember the actual parameters
```

Upsert takes a local lock so the new info is available immediately, UpsertToKVStore would call the listeners eventually. I don't know how sensitive we are to delays, though.

File: pkg/maps/lxcmap/lxcmap.go (hunk 3 above)
Comment (raybejjani) [reply pkg/maps/lxcmap/lxcmap.go]: I third this.

File: daemon/daemon.go [hunk 8]
```diff
@@ -905,6 +883,51 @@ func (d *Daemon) init() error {
 	return nil
 }
 
+// startLXCMapSync spawns a controller which synchronizes the host entries in BPF lxc Map
+// with the in-memory host IP address values.
+func (d *Daemon) startLXCMapSync() error {
```
Comment (raybejjani) [inline daemon/daemon.go]: Since this is a bit of a hack-fix, could you mention that here? Ideally we can ditch the controller one day.

File: pkg/bpf/endpoint.go (hunk 5 above)
Comment (raybejjani) [reply pkg/bpf/endpoint.go]: lbmap has a similar bug, if you end up making a separate PR to fix this.

**Review by raybejjani — APPROVED** [review]
the 1-line change to endpoint is approved. The other comments are just suggestions

File: pkg/bpf/endpoint.go (hunk 5 above)
Comment (jrajahalme) [reply pkg/bpf/endpoint.go]: @raybejjani Next commit fixes the value receiver for lbmap, but I just noted that I did not fix it for EndpointKey on which the wrong comment was on here. So the comment is wrong, and this too returns a pointer to the temporary. Fixing this forces all other methods on EndpointKey to also use pointer receiver, which is correct as well, as what is a point of taking a copy and then returning pointer to it. That fix will likely cascade further than we want for the bug fix, so I'll not touch it for now.

File: daemon/daemon.go (hunk 6 above)
Comment (jrajahalme) [reply daemon/daemon.go]: I'll do the first, as it is less complicated.

File: daemon/daemon.go (hunk 8 above)
Comment (jrajahalme) [reply daemon/daemon.go]: Will add a comment.

**Review by jrajahalme — COMMENTED** [review]
@tgraf lookups on the lxcmap depend on the GetValuePtr to have a pointer receiver, so there is a dependency between the commits. Do you mean that I should split this up to 3 separate PRs and then we backport the latter 2, in the order of dependency?

File: daemon/daemon.go (hunk 7 above)
Comment (joestringer) [reply daemon/daemon.go]: The reserved identities should specifically *not* be upserted to the kvstore. Host is semantically only the host namespace where the pod is running, not other nodes in the cluster.

File: pkg/maps/lxcmap/lxcmap.go [hunk 9]
```diff
@@ -192,10 +193,20 @@ func WriteEndpoint(f EndpointFrontend) error {
 // AddHostEntry adds a special endpoint which represents the local host
 func AddHostEntry(ip net.IP) error {
 	key := NewEndpointKey(ip)
-	ep := EndpointInfo{Flags: EndpointFlagHost}
+	ep := &EndpointInfo{Flags: EndpointFlagHost}
 	return LXCMap.Update(key, ep)
 }
 
+func SyncHostEntry(ip net.IP) error {
```
[the remark this answers is not in the document]
Comment (jrajahalme) [reply pkg/maps/lxcmap/lxcmap.go]: Fixed.

File: daemon/daemon.go (hunk 2 above)
Comment (jrajahalme) [reply daemon/daemon.go]: `OnIPIdentityCacheGC()` deletes stale entries from ipcache bpf map. Here we are adding potentially missing entries. I agree that these should be combined into a more comprehensive desired/realized state implementation, but that is out of scope for this PR.

File: pkg/maps/lxcmap/lxcmap.go (hunk 3 above)
Comment (jrajahalme) [reply pkg/maps/lxcmap/lxcmap.go]: Done.

**Comment by jrajahalme** [comment]
Carved out the pointer receiver fix to a separate PR: https://github.com/cilium/cilium/pull/3922. This one depends on that one being merged first, as I've had to cherry-pick it here as well. Reviews in this PR should be limited to the second commit only!

**Comment by jrajahalme** [comment]
test-me-please

File: daemon/daemon.go [hunk 10]
```diff
@@ -905,6 +899,43 @@ func (d *Daemon) init() error {
 	return nil
 }
 
+// syncLXCMap adds local host enties to bpf lxcmap, as well as
+// ipcache, if needed, and also notifies the daemon and network policy
+// hosts cache if changes were made.
+func (d *Daemon) syncLXCMap() error {
+	// TODO: Update addresses first, in case node addressing has changed.
+	// TODO: Once these start changing on runtime, figure out the locking strategy.
```
Comment (joestringer) [inline daemon/daemon.go:907]: What specifically are we deferring figuring out here? Locking of `lxcmap`? of `IPIdentityCache`?

File: daemon/daemon.go [hunk 11]
```diff
@@ -905,6 +899,43 @@ func (d *Daemon) init() error {
 	return nil
 }
 
+// syncLXCMap adds local host enties to bpf lxcmap, as well as
+// ipcache, if needed, and also notifies the daemon and network policy
+// hosts cache if changes were made.
+func (d *Daemon) syncLXCMap() error {
+	// TODO: Update addresses first, in case node addressing has changed.
```
Comment (joestringer) [inline daemon/daemon.go:906]: What is this todo and what are the implications of not doing it now?

**Review by joestringer — APPROVED** [review]
Seems fine to me, so long as the todos aren't pressing.

**Comment by jrajahalme** [comment]
test-me-please

File: daemon/daemon.go (hunk 11 above)
Comment (jrajahalme) [reply daemon/daemon.go:906]: If host device configuration (e.g., IP addressing) changes we do not reflect that change to the bpf maps, i.e., things will break.

File: daemon/daemon.go (hunk 10 above)
Comment (jrajahalme) [reply daemon/daemon.go:907]: All of it, I guess.

**Review by tgraf — APPROVED** [review]

File: daemon/daemon.go (hunk 7 above)
Comment (raybejjani) [reply daemon/daemon.go]: Oops. You are most correct. I'm still a little uneasy about invoking the listeners directly but this is the same init-time one-off we had so nothing really changed, I guess.
